### DISCUSSION TITLE
Model fails silently when saving unless 'saved' method exists in custom model

### DISCUSSION THREAD

**Discussion by fernando2amigos** [spatie/laravel-permission#1984] [Q&A]
I'm upgrading to PHP 8.1 (from 7.4) and with that also bumping versions of Laravel to 8.77.1 and of Laravel-Permissions 5.4.2

When running the migrations it simply stops after creating (saving to DB) the very first Permission or Role. 
After retrying, the just created Permission or Role goes fine then halts on the next one.
I've observed that the insert to DB has succeeded, but code stopped without throwing Query or general Exception, nor SQL error.

After debugging I've found that the issue is when the Laravel calls the model's `saved()` method:
```
# vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php

 protected function finishSave(array $options)
    {
        $this->fireModelEvent('saved', false); // <<<<< HERE IT FAILS

        if ($this->isDirty() && ($options['touch'] ?? true)) {
            $this->touchOwners();
        }

        $this->syncOriginal();
    }
```

This is our custom Permission model:
```
<?php

namespace Modules\IdentityAndAccess\Models;

use Ramsey\Uuid\Uuid;

/**
 * Class Permission
 * @package Modules\IdentityAndAccess\Models
 * @property string $name
 */
class Permission extends \Spatie\Permission\Models\Permission
{
    protected static function boot()
    {
        parent::boot();

        static::creating(function ($model) {
            $model->keyType = 'string';
            $model->incrementing = false;

            $model->{$model->getKeyName()} = $model->{$model->getKeyName()} ?: (string) Uuid::uuid4();
        });
    }

    public function getIncrementing(): bool
    {
        return false;
    }

    public function getKeyType(): string
    {
        return 'string';
    }

    /**
     * @inheritdoc
     */
    public static function saved($callback)
    {
        /* ... */ // If this method is not present then it fails
    }
}
```
As you can see, it extends Spatie's Permission model, which also extends Laravel's Model which also uses HasEvents trait which has the `saved()` method that in this case is overloaded by our custom one.
  
So I'm wondering the cause of this behavior: is it our code or maybe a bug in between Laravel and the package.  

Adding the `saved()` method to Permission allowed me to go further until I hit the same situation for Role model.

**Comment by erikn69** [comment]
There is no full support for `uuid`, #1909, it was working for you with `uuid`?

**Reply by fernando2amigos** [reply]
Yes, it was working so far (PHP 7.4 + Laravel 8.73.1 + Laravel-Permission 3.18).
Maybe it was because of v3.x instead of 5.x 
I'm moving forward and fixing other stuff but at least migrations and Users seeders ran so I feel it will work.
I'll keep you updated.
Thanks for your collaboration!
